### ISSUE TITLE
Error with Google Tokens

### ISSUE THREAD

**Issue by kayvanaarssen** [imapsync/imapsync#418]
Cannot seem to get the correct google Oauth tokens what can this be.
The file is there...

```bash
root@misc-public:~/sync/oauth2# ./generate_gmail_token USERNAME@DOMAIN.COM
I am PID 155746 ./generate_gmail_token USERNAME@DOMAIN.COM
In refresh_token_is_here USERNAME@DOMAIN.COM 
./D_oauth2_refresh_token_USERNAME@DOMAIN.COM.txt is not here or empty
Interractive
I am the parent
will do: script -q -f -c ./generate_gmail_token USERNAME@DOMAIN.COM 
I am PID 155751 ./generate_gmail_token USERNAME@DOMAIN.COM
In refresh_token_is_here USERNAME@DOMAIN.COM 
./D_oauth2_refresh_token_USERNAME@DOMAIN.COM.txt is not here or empty
Interractive
I am a child in script 155751
./generate_gmail_token: 28: /root/sync/oauth2/oauth2_google.py: not found
refresh_token saved in ./D_oauth2_refresh_token_USERNAME@DOMAIN.COM.txt
access_token saved in ./D_oauth2_access_token_USERNAME@DOMAIN.COM.txt
/root/sync/oauth2/oauth2_google.py --quiet --user=USERNAME@DOMAIN.COM --client_id=108687549524-86sjq07f3ch8otl9fnr56mjnniltdrvn.apps.googleusercontent.com --client_secret=zAJO4PLxzeJ4yOaiJRk6f69k --refresh_token=
./generate_gmail_token: 52: /root/sync/oauth2/oauth2_google.py: not found
access_token saved in ./D_oauth2_access_token_USERNAME@DOMAIN.COM.txt
./generate_gmail_token: 89: /root/sync/oauth2/oauth2_google.py: not found
oauth2_string saved in ./D_oauth2_oauthdirect_USERNAME@DOMAIN.COM.txt
Give this to imapsync --oauthdirect1 or --oauthdirect2
'```

**Comment by gilleslamiral** [comment]
Do you have 
https://imapsync.lamiral.info/oauth2/oauth2_gmail/oauth2_google.py
in the same directory?

**Comment by kayvanaarssen** [comment]
Yes, just did a whole clone of the IMAPSYNC from github as well.

See:
![image](https://github.com/imapsync/imapsync/assets/672841/88ec7252-4f09-4f4e-b0a8-420df9e1c1dc)

**Comment by gilleslamiral** [comment]
It's one of those shell mysteries...

What gives 
`/root/sync/oauth2/oauth2_google.py`

**Comment by kayvanaarssen** [comment]
Looks like it cannot find Phyton.

```
root@misc-public:~/sync/oauth2# ./oauth2_google.py
-bash: ./oauth2_google.py: /usr/bin/python: bad interpreter: No such file or directory
root@misc-public:~/sync/oauth2#
```

**Comment by kayvanaarssen** [comment]
@gilleslamiral I've started to try it this morning when I generate with: 

```bash
./oauth2_google.py --user=USER@DOMAIN.COM --client_id=6395189-hbl1cs6k98a2aum.apps.googleusercontent.com     --client_secret=GOCSL_sEckDXEFx     --generate_oauth2_token
```

And i copy and past the URL your script gives me i get this error in google: 

![image](https://github.com/imapsync/imapsync/assets/672841/9e2c2b8e-118e-4055-8479-d2edc63dd1ac)

Where can I find the right guide on how to test this, so that we do not need to enable 2FA when doing a migration and creating App Passwords.

We do Many M365 Migrations that works great but Google is a pain (at the moment :-D )

**Comment by gilleslamiral** [comment]
The script to run is `./generate_gmail_token` not `./oauth2_google.py`

**Comment by kayvanaarssen** [comment]
than i get this error on Google:

![image](https://github.com/imapsync/imapsync/assets/672841/85a868fb-4890-4731-87b1-27a63d84777b)

**Comment by gilleslamiral** [comment]
You may use an old release.

Update with:
https://imapsync.lamiral.info/oauth2/oauth2_gmail/

**Comment by kayvanaarssen** [comment]
@gilleslamiral Still having the same issue just tried again:

![image](https://github.com/imapsync/imapsync/assets/672841/47d7350b-ecc3-4d92-b289-6b3c1dd30c70)

![image](https://github.com/imapsync/imapsync/assets/672841/ec334106-68cb-476d-820d-25b5a521de97)

**Comment by gilleslamiral** [comment]
Do you use `generate_gmail_token  `release 1.12?
````
gilles@petite:~/public_html/imapsync/oauth2/oauth2_gmail 11$ more generate_gmail_token
#!/bin/sh

# $Id: generate_gmail_token,v 1.12 2023/05/31 09:31:42 gilles Exp gilles $
````

https://imapsync.lamiral.info/oauth2/oauth2_gmail/generate_gmail_token

**Comment by kayvanaarssen** [comment]
Yes I do, downloaded from the link you provided: 

![image](https://github.com/imapsync/imapsync/assets/672841/061c8103-ebcd-4243-bf08-91c1b39af924)

**Comment by gilleslamiral** [comment]
Ok
I don't have this reject when I use it, I just have warnings I can accept

**Comment by robertmathews** [comment]
I have the same problem: when I try using `generate_gmail_token,v 1.12` with Google, I get the `Access blocked: imapsync has not completed the Google verification process` error:
<img width="469" alt="Screenshot 2024-02-22 at 9 41 23 AM" src="https://github.com/imapsync/imapsync/assets/14928992/fae92e71-7838-4ee2-9f02-90e5eb894ce0">

According to [the Google documentation](https://developers.google.com/identity/protocols/oauth2/production-readiness/brand-verification?hl=en#projects-used-in-dev-test-stage), they allow a certain number of users to be shown a warning they can click past, but when you reach a "user cap", everyone else gets blocked:

>Note: Your app is still subject to a tester warning screen, as shown in figure 2, a user cap is in effect, and the refresh token lifetime is limited.

So this is not going to work for other people unless you follow the [How to submit your app for verification](https://developers.google.com/identity/protocols/oauth2/production-readiness/brand-verification?hl=en#submit-app-for-verification) steps.

**Comment by gilleslamiral** [comment]
Ok.
Maybe you can try with the Thunderbird client_id

**Comment by Avamander** [comment]
Google has now blocked all OOB OAuth2 flows:

```
The out-of-band (OOB) flow has been blocked in order to keep users secure. Follow the Out-of-Band (OOB) flow migration guide linked in the developer docs below to migrate your app to an alternative method.
Request details: redirect_uri=urn:ietf:wg:oauth:2.0:oob
```

The current scripts unfortunately can't function even if you use your own Client ID and Secret.

**Comment by Avamander** [comment]
The script `generate_gmail_token` doesn't use OOB anymore. But the script [`oauth2_google.py`](https://imapsync.lamiral.info/oauth2/oauth2_gmail/oauth2_google.py) invoked by `generate_gmail_token` defines `REDIRECT_URI = 'urn:ietf:wg:oauth:2.0:oob'`.

I'm unfortunately not sure what the replacement should be though, the script would have to start a HTTP listener for that at localhost, no?

**Comment by gilleslamiral** [comment]
> The script `generate_gmail_token` doesn't use OOB anymore. But the script [`oauth2_google.py`](https://imapsync.lamiral.info/oauth2/oauth2_gmail/oauth2_google.py) invoked by `generate_gmail_token` defines `REDIRECT_URI = 'urn:ietf:wg:oauth:2.0:oob'`.
> 

So I was wrong about OOB. I don't understand why this still works for me.

> I'm unfortunately not sure what the replacement should be though, the script would have to start a HTTP listener for that at localhost, no?

Yes. I wrote this for Office365:
https://imapsync.lamiral.info/oauth2/oauth2_office365/oauth2_office365_with_imap

It should work for Gmail, changing the values. I haven't trie it.
